### ISSUE TITLE
accuracy never change

### ISSUE THREAD

**Issue by sodaling** [cerndb/dist-keras#9]
My code and dataset likes this:
https://github.com/sodaling/embedding_test
Can  you help me  test my code? Because no matter how I modify the code or trainer,the  accuracy often 50%.(The dataset's label is 1 or 0,trainer model's predictions are all nearly 1 or 0.  )

**Comment by JoeriHermans** [comment]
Have you tried to increase the number of epochs? I'll test your code tomorrow, so I can have a better look what is going on.

Joeri

**Comment by sodaling** [comment]
I have tried with different epochs or different trainer.They all came out with all nearly 0 or nearly 1.So the accuracy  always equal to 0.5(The training dataset consists of half-tag 1 and half-tag 0.)

**Comment by JoeriHermans** [comment]
The values in your dataset, what do they "mean"? Are they indices? Because I have the feeling that might be your issue.

**Comment by sodaling** [comment]
The number of my dataset is a set of id's for each of my words.Each sentence's word is replaced by the word's id.Then the embedding layer accept it and turn it into new vector for lstm layer.
But i think it isn't the point of my bug.I have tested my local version code of this model and ran it successfully.......

**Comment by sodaling** [comment]
I think i have figured  out why it happened.I made some test during this time.Such as : 

1,Changing  your mnist code by removing the label 2 to 9 data.It left 0 and 1 label and features.And removed the MinMaxTransformer,changed loss_mlp to binary_crossentropy.So your mnist code like my lstm code.And afer i trainse it , the prediction is all 1 even though half of it are 0.It like my lstm code's output.

2,Chaging my model to simple Dense model not lstm and embedding model like your mnist code.It's accuracy never change again.The prediction  is always 1 or always 0.I added MinMaxTransformer like but it didn't worked.I check the features_normalized and found it all likes "4.3999999999988493E-4".I think the model's features like this can't be trained.

My conclusion is ,when the training data is to big or to small.The distribute model can't work like in single computer.Hope for your reply.Thanks a lots.

**Comment by JoeriHermans** [comment]
Sorry, for the late reply, I was on a holiday.

Are you using SingleTrainer as a _trainer_ (as in the sample you provided)? Then it should produce the same behavior as local training since this procedure doesn't apply any distributed training mechanism. If this is the case, I would say it is a problem with the way how you specify the features to the embedding.

However, if this is not the case, I would say that it probably is due to the parallelization of gradient descent. Not that this is still an active field of study because of the reasons I have described here https://db-blog.web.cern.ch/blog/joeri-hermans/2017-01-distributed-deep-learning-apache-spark-and-keras Especially regarding _"implicit momentum"_.

Could you update the code in your GitHub repository, so I could check your new implementation?

**Comment by sodaling** [comment]
I train my model with any trainer you provided,but still failed.I will update my new implementation soon.It's dataset based on your mnist dataset except that the data only retains 0 and 1.

**Comment by JoeriHermans** [comment]
If the accuracy doesn't increase when using ```SingleTrainer```, I would say you are doing something wrong in the modeling. But ok, I'll keep track of your repo and test it.

**Comment by sodaling** [comment]
https://github.com/sodaling/embedding_test/blob/master/embeding2.py
Here is my new implementation.
The accuracy doesn't increase when using all trainer......In my new implementation Whatever the dataset's label is 0 or 1,the prediction is always 1.......

**Comment by JoeriHermans** [comment]
Yes, that is what I'm saying. If even ```SingleTrainer``` doesn't work, the cause is likely in your modeling. I'll try it out now.

**Comment by sodaling** [comment]
Waiting for your good news and thanks....

**Comment by sodaling** [comment]
If i modified your mnist's code from Convolution2D to Convolution1D,and change the dataset to 784*1 matrix:
```python
reshape_transformer = ReshapeTransformer("features_dense", "matrix1", (784, 1))
dataset = reshape_transformer.transform(dataset)
model1 = Sequential()
model1.add(GlobalMaxPooling1D())
model1.add(Dense(hidden_dims))
model1.add(Dropout(0.2))
model1.add(Activation('relu'))
model1.add(Dropout(0.5))
model1.add(Dense(1))
optimizer_mlp = 'adam'
mlp.add(Activation('sigmoid'))
loss_mlp = 'binary_crossentropy'
```
p.s:The dataset only retains 0 and 1 features and label.
The prediction is almost all 0.5 even the label are 0 or 1.But GlobalMaxPooling1D is okay likes your mnist code.

**Comment by sodaling** [comment]
+--------------------+-----+
|prediction          |label|
+--------------------+-----+
|[0.5279000401496887]|1    |
|[0.5279000401496887]|0    |
|[0.5279000401496887]|1    |
|[0.5279000401496887]|0    |
|[0.5279000401496887]|0    |
|[0.5279000401496887]|1    |
|[0.5279000401496887]|1    |
|[0.5279000401496887]|0    |
|[0.5279000401496887]|0    |
|[0.5279000401496887]|1    |
|[0.5279000401496887]|1    |
|[0.5279000401496887]|1    |
|[0.5279000401496887]|1    |
|[0.5279000401496887]|1    |
|[0.5279000401496887]|0    |
|[0.5279000401496887]|1    |
|[0.5279000401496887]|1    |
|[0.5279000401496887]|1    |
|[0.5279000401496887]|0    |
|[0.5279000401496887]|1    |
+--------------------+-----+
only showing top 20 rows

**Comment by sodaling** [comment]
I think i have figured it out!
There are two code.The first code is 

```python

......
reshape_transformer = ReshapeTransformer("features_dense", "matrix1", (784, 1))
dataset = reshape_transformer.transform(dataset)
model1 = Sequential()
model1.add(LSTM(128,input_shape=(784,1)))
model1.add(Dropout(0.5))
model1.add(Dense(1))
model1.add(Activation('sigmoid'))
......
```
It's trained prediction is 
```
+--------------------+-----+
|prediction          |label|
+--------------------+-----+
|[0.5194315314292908]|1    |
|[0.519431471824646] |0    |
|[0.5194315314292908]|1    |
|[0.5194315314292908]|0    |
|[0.5194315314292908]|0    |
|[0.519431471824646] |1    |
|[0.519431471824646] |1    |
|[0.5194315314292908]|0    |
|[0.5194315314292908]|0    |
|[0.5194315314292908]|1    |
|[0.5194315314292908]|1    |
|[0.5194315314292908]|1    |
|[0.5194315314292908]|1    |
|[0.5194315314292908]|1    |
|[0.5194315314292908]|0    |
|[0.519431471824646] |1    |
|[0.5194315314292908]|1    |
|[0.5194315314292908]|1    |
|[0.5194315314292908]|0    |
|[0.5194315314292908]|1    |
+--------------------+-----+

```

The second code is:
```python
reshape_transformer = ReshapeTransformer("features_dense", "matrix2", (28, 28))
dataset = reshape_transformer.transform(dataset)
model1 = Sequential()
model1.add(LSTM(128,input_shape=(28,28)))
model1.add(Dropout(0.5))
model1.add(Dense(1))
model1.add(Activation('sigmoid'))
model1.summary()

```
It's trained prediction is 
```
|prediction             |label|
+-----------------------+-----+
|[0.9999724626541138]   |1    |
|[8.034089114516973E-5] |0    |
|[0.9999895095825195]   |1    |
|[1.312557578785345E-4] |0    |
|[4.710748908109963E-4] |0    |
|[0.9999842643737793]   |1    |
|[0.9999681711196899]   |1    |
|[1.910389109980315E-4] |0    |
|[1.4313575229607522E-4]|0    |
|[0.9999898672103882]   |1    |
|[0.9999815821647644]   |1    |
|[0.9999613761901855]   |1    |
|[0.999988317489624]    |1    |
|[0.9999874234199524]   |1    |
|[1.5818639076314867E-4]|0    |
|[0.9999778866767883]   |1    |
|[0.9999799728393555]   |1    |
|[0.9999681711196899]   |1    |
|[6.574992585228756E-5] |0    |
|[0.9999677538871765]   |1    |

```
Obviously,the first one succeed but the second didn't.I think maybe it's the  reason of all question.Could you tell me why it happened?So i can modified my embedding code.

**Comment by JoeriHermans** [comment]
It's clearly a modeling issue. The first example you are 784 examples of 1 pixel. Whereas, in the second example, you are providing 28 examples of 28 pixels. You can view it like this.

Imagine having a time-series where you supply a certain window to your LSTM. If your window is small, you probably cannot tell what will happen in the future, because you use a small piece of information to make predictions from. However, if you have a large window (like your second example), you probably will make better predictions, because you have more information. 

For example, if you would like to predict the weather of next week, using only a minute of temperature data, will your predictions be accurate?

**Comment by sodaling** [comment]
As you explained,i change the model and dataset's shape to (784,1)，and it predicted sucessfully.
So i think the error of my embedding is the input shape problem.It take all the input vector in 1 pixels.It won't happen  while local training without Spark's dataframe input.
I will proved it later and thanks.

**Comment by JoeriHermans** [comment]
Just to clarify, the following model

```python
reshape_transformer = ReshapeTransformer("features_dense", "matrix2", (28, 28))
dataset = reshape_transformer.transform(dataset)
model1 = Sequential()
model1.add(LSTM(128,input_shape=(28,28)))
model1.add(Dropout(0.5))
model1.add(Dense(1))
model1.add(Activation('sigmoid'))
model1.summary()
```

worked correctly right? Because its predictions are not random.

**Comment by sodaling** [comment]
My faust,I confused them.Let's clarify it.
(784, 1) worked failed.
(1,784) and (28,28) worked sucessfully.

**Comment by JoeriHermans** [comment]
Thanks, that makes sense :D

**Comment by sodaling** [comment]
But i still can't run my embedding code sucessfully.
```python
reshape_transformer = ReshapeTransformer("features_dense", "matrix4", (784))
dataset = reshape_transformer.transform(dataset)
model1 = Sequential()
model1.add(Embedding(input_dim=256, output_dim=256))
model1.add(LSTM(128))
model1.add(Dropout(0.5))
model1.add(Dense(1))
model1.add(Activation('sigmoid'))
```
And it's prediction is 
```
+--------------------+-----+
|prediction          |label|
+--------------------+-----+
|[0.5419784784317017]|1    |
|[0.5419784784317017]|0    |
|[0.5419784784317017]|1    |
|[0.5419784784317017]|0    |
|[0.5419784784317017]|0    |
|[0.5419784784317017]|1    |
|[0.5419784784317017]|1    |
|[0.5419784784317017]|0    |
|[0.5419784784317017]|0    |
|[0.5419784784317017]|1    |
|[0.5419784784317017]|1    |
|[0.5419784784317017]|1    |
|[0.5419784784317017]|1    |
|[0.5419784784317017]|1    |
|[0.5419784784317017]|0    |
|[0.5419784784317017]|1    |
|[0.5419784784317017]|1    |
|[0.5419784784317017]|1    |
|[0.5419784784317017]|0    |
|[0.5419784784317017]|1    |
+--------------------+-----+


```
It can't set embedding's shape like lstm layer.And the prediction are all the same not like (784, 1) test.So the issue is still on....

**Comment by sodaling** [comment]
Finally , i found a way to solve it.
I change my embedding and LSTM model to :
```python
model1 = Sequential()
model1.add(Embedding(input_dim=52965, output_dim=1, dropout=0.2,input_length=50))
model1.add(Flatten())
model1.add(Reshape((1,50)))
model1.add(LSTM(128))
model1.add(Dropout(0.5))
model1.add(Dense(1))
model1.add(Activation('sigmoid'))
```
Then it worked.I still don't know why.Because it is different from the model i used on single machine:
```python
model1 = Sequential()
model1.add(Embedding(input_dim=52965, output_dim=256))
model1.add(LSTM(128))
model1.add(Dropout(0.5))
model1.add(Dense(1))
model1.add(Activation('sigmoid'))
```

**Comment by sodaling** [comment]
Whatever,the issue about embedding and LSTM model still unresolved.Hope anyone can figure it out.

**Comment by JoeriHermans** [comment]
I really think this is a modeling issue, or you didn't sort the time series on the Spark partitions.

**Comment by raviolli** [comment]
No one figured this out?

**Comment by anishsharma** [comment]
Has any one figure this out ? After training all my model weights are around 0.5
